openpgp/write.go

```diff
@@ -60,19 +60,25 @@ func armoredDetachSign(w io.Writer, signer *Entity, message io.Reader, sigType p
 }
 
 func detachSign(w io.Writer, signer *Entity, message io.Reader, sigType packet.SignatureType, config *packet.Config) (err error) {
-	if signer.PrivateKey == nil {
-		return errors.InvalidArgumentError("signing key doesn't have a private key")
+	signingKey, ok := signer.SigningKey(config.Now())
+	if !ok {
+		return errors.InvalidArgumentError("no valid signing keys")
+	}
+	if signingKey.PrivateKey == nil {
+		return errors.InvalidArgumentError("signing subkey doesn't have a private key")
 	}
-	if signer.PrivateKey.Encrypted {
-		return errors.InvalidArgumentError("signing key is encrypted")
+	if signingKey.PrivateKey.Encrypted {
+		return errors.InvalidArgumentError("signing subkey is encrypted")
```

Comment:
Since it's still possible for `signer.SigningKey` to return the primary key (if there is no signing subkey) I would prefer to keep the two error messages as they were, saying "signing key".